File: cpp/behavior/observer.cpp
#include <iostream>
#include <list>
using namespace std;

class Observer {
public:
    virtual void update() = 0;
};

class ConcreteObserver1: public Observer {
public:
    void update() {
        cout<<"ConcreteObserver1 update" <<endl;
    }
};
class ConcreteObserver2: public Observer {
public:
    void update() {
        cout<<"ConcreteObserver2 update" <<endl;
    }
};

class Subject {
private:
    list<Observer*> obs;
public:
    void regist (Observer* o) {
        obs.push_back(o);
    }
    void update() {
        for (list<Observer*>::iterator iter = obs.begin(); iter != obs.end(); ++iter) {
            (*iter)->update();
        }
    }
};

int main() {
    Subject s;
    s.regist(new ConcreteObserver1());
    s.regist(new ConcreteObserver2());
    s.regist(new ConcreteObserver1());
    s.update();
    return 0;
}

File: golang/behavior/observer.go
package main

import "container/list"

type Observer interface {
	update()
}
type ConcreteObserver1 struct {

}
func (o *ConcreteObserver1) update () {
	println("ConcreteObserver1 update")
}
type ConcreteObserver2 struct {

}
func (o *ConcreteObserver2) update () {
	println("ConcreteObserver2 update")
}

type Subject struct {
	obs *list.List
}
func NewSubject() *Subject {
	return &Subject{list.New()}
}
func (s *Subject) regist(o Observer) {
	s.obs.PushBack(o)
}
func (s *Subject) update() {
	for e := s.obs.Front(); e != nil; e = e.Next() {
		e.Value.(Observer).update()
	}
}

func main() {
	s := NewSubject()
	s.regist(&ConcreteObserver1{})
	s.regist(&ConcreteObserver2{})
	s.regist(&ConcreteObserver1{})
	s.update()
}